src/main/java/org/jenkinsci/plugins/workflow/support/steps/input/InputAction.java

```diff
@@ -46,18 +64,27 @@ public void onLoad(Run<?, ?> r) {
                 // Loading from before JENKINS-25889 fix. Load the IDs and discard the executions, which lack state anyway.
                 assert executions != null && !executions.contains(null) : executions;
                 ids = new ArrayList<String>();
+                if (inputDefinitions == null) {
+                    inputDefinitions = new HashMap<String, InputPromptDefinition>();
+                }
                 for (InputStepExecution execution : executions) {
                     ids.add(execution.getId());
+                    inputDefinitions.put(execution.getId(), new InputPromptDefinition(execution.input));
                 }
                 executions = null;
             }
         }
     }
 
     @SuppressFBWarnings(value="EC_UNRELATED_TYPES_USING_POINTER_EQUALITY", justification="WorkflowRun implements Queue.Executable")
-    private synchronized void loadExecutions() {
+    private synchronized void loadExecutions() throws InterruptedException, TimeoutException {
         if (executions == null) {
+            boolean needsSave = false;
             executions = new ArrayList<InputStepExecution>();
+            if (inputDefinitions == null) {
+                inputDefinitions = new HashMap<String, InputPromptDefinition>();
```

Comment:
Generally prefer to use `TreeMap` (or `LinkedHashMap`) for more deterministic behavior.